### PR TITLE
refactor(bridges): Factor out approval functions

### PR DESCRIPTION
**Pull request by pxrl**
For future reuse by a slimmed-down bridge adapter structure.

### CODE REVIEW

File: src/clients/bridges/BaseAdapter.ts [hunk 1]
```diff
@@ -113,88 +110,67 @@ export abstract class BaseAdapter {
...
+        return allowance;
       })
     );
 
-    await utils.forEachAsync(allowances, async (allowance, index) => {
-      if (allowance && allowance.lt(toBN(MAX_SAFE_ALLOWANCE))) {
-        tokensToApprove.push({ l1Token: l1TokenContracts[index], targetContract: associatedL1Bridges[index] });
-      } else {
-        // Save allowance in cache with no TTL as these should never decrement.
-        if (redis) {
```
Comment (pxrl) [inline src/clients/bridges/BaseAdapter.ts:165]: Incidental optimisation - this looks like it was setting a record in redis even if we already read the allowance from redis. Removing it should save quite a few writes to redis on update.

**Review by dohaki — APPROVED** [review]
LGTM

File: src/clients/bridges/BaseAdapter.ts [hunk 2]
```diff
@@ -113,88 +111,63 @@ export abstract class BaseAdapter {
...
+          return allowance;
         }
+
+        // If the onchain allowance is > MAX_SAFE_ALLOWANCE, cache it for next time.
+        allowance = await l1Token.allowance(address, l1Bridge);
+        if (allowance.gte(MAX_SAFE_ALLOWANCE)) {
+          await setTokenAllowanceInCache(l1Token.address, address, l1Bridge, allowance);
+        }
+
+        return allowance;
```
Comment (james-a-morris) [inline src/clients/bridges/BaseAdapter.ts:152]: We may be able to reduce the diff if we wrapped this 
```suggestion
        if(!allowance) {
          // If the onchain allowance is > MAX_SAFE_ALLOWANCE, cache it for next time.
          allowance = await l1Token.allowance(address, l1Bridge);
          if (allowance.gte(MAX_SAFE_ALLOWANCE)) {
            await setTokenAllowanceInCache(l1Token.address, address, l1Bridge, allowance);
          }
        }
        return allowance;
```

**Review by james-a-morris — COMMENTED** [review]
one nit - good call to move this out of scope into a utils directory

File: src/clients/bridges/BaseAdapter.ts (hunk 2 above)
Comment (pxrl) [reply src/clients/bridges/BaseAdapter.ts]: Good call - implemented: https://github.com/across-protocol/relayer/pull/1620/commits/9ee231677a4d97ede40ed3f83f2aa60ad99dede3

**Review by bmzig — APPROVED** [review]

**Review by james-a-morris — APPROVED** [review]

**Pull request merged**